src/types/tdigest.cc

```diff
@@ -0,0 +1,439 @@
+      }
+      --ci_left;
+      diff += td[ci_left].weight / 2 + td[ci_right].weight / 2;
+    }
+
+    // interpolate from adjacent centroids
+    diff /= (td[ci_left].weight / 2 + td[ci_right].weight / 2);
+    return Lerp(td[ci_left].mean, td[ci_right].mean, diff);
+  }
+
+  double Mean() const {
+    double sum = 0;
+    for (const auto& centroid : tdigests_[current_]) {
+      sum += centroid.mean * centroid.weight;
+    }
+    return total_weight_ == 0 ? NAN : sum / total_weight_;
+  }
+
+  double TotalWeight() const { return total_weight_; }
+
+ private:
+  // must be declared before merger_, see constructor initialization list
+  const uint32_t delta_;
+
+  TDigestMerger<> merger_;
+  double total_weight_;
+  double min_, max_;
+
+  // ping-pong buffer holds two tdigests, size = 2 * delta * sizeof(Centroid)
+  std::vector<Centroid> tdigests_[2];
+  // index of active tdigest buffer, 0 or 1
+  int current_;
+};
+
+TDigest::TDigest(uint64_t delta) : impl_(std::make_unique<TDigestImpl>(delta)) { Reset({}); }
+
+void TDigest::Merge(const std::vector<TDigest>& others) {
+  if (others.empty()) {
+    return;
+  }
+
+  std::vector<const TDigestImpl*> impls;
+  impls.reserve(others.size());
+
+  std::transform(others.cbegin(), others.cend(), std::back_inserter(impls),
+                 [](const TDigest& i) { return i.impl_.get(); });
+
+  impl_->Merge(impls);
+}
+
+void TDigest::Reset(const CentroidsWithDelta& centroids_list) {
+  impl_->Reset(centroids_list.centroids, centroids_list.min, centroids_list.max, centroids_list.total_weight);
+}
+
+void TDigest::Reset() { impl_->Reset(); }
+
+CentroidsWithDelta TDigest::DumpCentroids() const {
+  auto centroids = impl_->Centroids();
+  return {
+      .centroids = std::move(centroids),
+      .delta = impl_->Delta(),
+      .min = impl_->Min(),
+      .max = impl_->Max(),
+      .total_weight = impl_->TotalWeight(),
+  };
+}
+
+void TDigest::Add(std::vector<double> items) { impl_->MergeInput(items); }
+
+StatusOr<CentroidsWithDelta> TDigestMerge(const std::vector<CentroidsWithDelta>& centroids_list) {
+  if (centroids_list.empty()) {
+    return Status{Status::InvalidArgument, "centroids_list is empty"};
+  }
+  if (centroids_list.size() == 1) {
+    return centroids_list.front();
+  }
+
+  TDigest digest{centroids_list.front().delta};
+  digest.Reset(centroids_list.front());
+
+  std::vector<TDigest> others;
+  others.reserve(centroids_list.size() - 1);
+
+  for (size_t i = 1; i < centroids_list.size(); ++i) {
+    TDigest d{centroids_list[i].delta};
+    digest.Reset(centroids_list[i]);
+    others.emplace_back(std::move(d));
+  }
+
+  digest.Merge(others);
+
+  return digest.DumpCentroids();
+}
+StatusOr<CentroidsWithDelta> TDigestMerge(const std::vector<double>& buffer, const CentroidsWithDelta& centroid_list) {
+  TDigest digest{centroid_list.delta};
+  digest.Reset(centroid_list);
+  digest.Add(buffer);
+  return digest.DumpCentroids();
+}
```

Comment:
We can leave an empty line in the end.

Reply:
Hi @PragmaTwice ,

Got it!

I have added a new endline. 😄

Best Regards,
Edward

Reply:
Would you mind to introduce a newline at end for every new source file?

Reply:
Of course!

I will check all files and add each file a newline at end. 😄

Reply:
Hi @PragmaTwice ,

I have added endline for each file in this PR. 😄
Please let me know if I missed any file.

Best Regards,
Edward